feat(curriculum): add string method workshop (#66936)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-employee-profile-validator/69e73a141addc6f442924b0d.md

---
id: 69e73a141addc6f442924b0d
title: Step 12
challengeType: 20
dashedName: step-12
---

# --description--

You already used `.find()` and slicing on the `email` variable. Now you will use the same technique on `clean_email`.

> Find the position of the `@` character in `clean_email` and store it in `at_pos`. Then use slicing to extract everything before `@` and store it in `email_user`.

Reviewer: The hint requires `clean_email.find('@')`, but the description only says "find the position", which `.index('@')` also does and which the hint rejects:

```python
Node("p = e.index('@')").find_variable('p').is_equivalent("p = e.find('@')")  # False
```

Step 3 names the method in its description, so naming it here keeps the graded requirement stated. "the same technique" on L10 gets a camper most of the way there, but it is worth being explicit.

```suggestion
Use `.find()` to get the position of the `@` character in `clean_email` and store it in a variable named `at_pos`. Then use slicing to extract everything before `@` and store it in a variable named `email_user`.
```

---
id: 69e73a141addc6f442924b0d
title: Step 12
challengeType: 20
dashedName: step-12
---

# --description--

You already used `.find()` and slicing on the `email` variable. Now you will use the same technique on `clean_email`.

Find the position of `'@'` in `clean_email` and store it in `at_pos`. Then use slicing to extract everything before `'@'` and store it in `email_user`.

> Then print `clean_email` with the prefix `'Email: '`. Also print `email_user` with the prefix `'Username: '`.

Reviewer 2: ```suggestion
Find the position of the `@` character in `clean_email` and store it in `at_pos`. Then use slicing to extract everything before `@` and store it in `email_user`.

Then print the result of concatenating the string `Email: ` and the `clean_email` variables. Also concatenate the string `Username: ` and the `email_user` variable and print the result.
```

> Also, print `email_user` with the lable `"Username: "`

Reviewer 3: ```suggestion
Find the position of `'@'` in `clean_email` and store it in `at_pos`. Then use slicing to extract everything before `'@'` and store it in `email_user`.

Then print `clean_email` with the label `'Email: '`. Also print `email_user` with the label `'Username: '`.
```

# --hints--

You should create a variable named `at_pos`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('at_pos')`)) })
```

> You should use `.find('@')` on `clean_email` to assign `at_pos` a value.

Reviewer 2: ```suggestion
You should call `.find('@')` on `clean_email` and assign the result to the `at_pos` variable.
```

```js
({ test: () => assert(runPython(`
node = _Node(_code).find_variable('at_pos')

> not node.is_empty() and node.is_equivalent("at_pos = clean_email.find('@')")

Reviewer 2: ```suggestion
_Node(_code).find_variable('at_pos').is_equivalent("at_pos = clean_email.find('@')")
```

`)) })
```

You should create a variable named `email_user`.

```js
({ test: () => assert(runPython(`_Node(_code).has_variable('email_user')`)) })
```

> You should use slicing with `at_pos` to assign `email_user` a value.

Reviewer: Names the string being sliced, as step 4's equivalent hint does.

```suggestion
You should use slicing with `at_pos` on `clean_email` to assign `email_user` a value.
```

> You should use slicing with `at_pos` on `clean_email` to assign `email_user` a value.

Reviewer 2: ```suggestion
You should use slicing with `at_pos` on `clean_email` and assign the result to the `email_user` variable.
```

```js
({ test: () => assert(runPython(`
node = _Node(_code).find_variable('email_user')

> not node.is_empty() and node.is_equivalent("email_user = clean_email[:at_pos]")

Reviewer 2: ```suggestion
node = _Node(_code).find_variable('email_user')
node.is_equivalent("email_user = clean_email[:at_pos]") or node.is_equivalent("email_user = clean_email[0:at_pos]")
```

`)) })
```

> You should print `clean_email` with the prefix `'Email: '`.

Reviewer 2: ```suggestion
You should print the result of concatenating the string `Email: ` and the `clean_email` variable.
```

```js
({ test: () => assert(runPython(`_Node(_code).has_call("print('Email: ' + clean_email)")`)) })
```

> You should print `email_user` with the prefix `'Username: '`.

Reviewer 2: ```suggestion
You should print the result of concatenating the string `Username: ` and the `email_user` variable.
```
